File: src/main/scala/org/joisen/chapter09/FakeWindowExample.scala
package org.joisen.chapter09

import org.apache.flink.api.common.state._
import org.apache.flink.streaming.api.functions.KeyedProcessFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector
import org.joisen.chapter05.{ClickSource, Event}

/**
 * @Author Joisen
 * @Date 2022/11/24 10:45
 * @Version 1.0
 */
object FakeWindowExample {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    env.addSource(new ClickSource)
      .assignAscendingTimestamps(_.timestamp)
      .keyBy(_.user)
      .process(new FakeWindow(10))
      .print()

    env.execute()

  }
  // 自定义的KeyedProcessFunction
  class FakeWindow(size: Long) extends KeyedProcessFunction[String, Event, String] {

    // 定义一个映射状态(key：窗口起始时间， value: 要保存的值)， 用来保存每个窗口的pv值
    lazy val windowPvMapState: MapState[Long, Long] = getRuntimeContext.getMapState(new MapStateDescriptor[Long, Long]("window-pv", classOf[Long], classOf[Long]))

    override def processElement(i: Event, context: KeyedProcessFunction[String, Event, String]#Context, collector: Collector[String]): Unit = {
      // 计算当前数据落入的窗口起始时间戳
      val start: Long = i.timestamp / size * size
      val end: Long = start + size
      // 注册一个定时器，用来触发窗口计算
      context.timerService().registerEventTimeTimer(end - 1)
      // 更新状态，count+1
      if(windowPvMapState.contains(start)){
        val count: Long = windowPvMapState.get(start)
        windowPvMapState.put(start, count+1)
      }else{
        windowPvMapState.put(start, 1L)
      }
    }

    override def onTimer(timestamp: Long, ctx: KeyedProcessFunction[String, Event, String]#OnTimerContext, out: Collector[String]): Unit = {
      // 定时器触发时，窗口输出结果
      val start: Long = timestamp + 1 - size
      val pv: Long = windowPvMapState.get(start)
      out.collect(s"url: ${ctx.getCurrentKey} 浏览量为：${pv}  窗口为：${start} ~ ${start + size}")
      // 窗口销毁
      windowPvMapState.remove(start)

    }
  }

}

File: src/main/java/org/joisen/java/chapter09/FakeWindowExample.java
package org.joisen.java.chapter09;

import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.state.MapState;
import org.apache.flink.api.common.state.MapStateDescriptor;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;
import org.joisen.java.chapter05.ClickSource;
import org.joisen.java.chapter05.Event;

import java.sql.Timestamp;
import java.time.Duration;

/**
 * @Author Joisen
 * @Date 2022/12/9 15:21
 * @Version 1.0
 */
public class FakeWindowExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        SingleOutputStreamOperator<Event> stream = env.addSource(new ClickSource())
                // 乱序流的watermark生成
                .assignTimestampsAndWatermarks(WatermarkStrategy.<Event>forBoundedOutOfOrderness(Duration.ZERO)
                        .withTimestampAssigner(new SerializableTimestampAssigner<Event>() {
                            @Override
                            public long extractTimestamp(Event event, long l) {
                                return event.timestamp;
                            }
                        }));
        stream.print("input ");

        stream.keyBy(data -> data.url)
                        .process(new FakeWindowResult(10000L))
                                .print();

        env.execute();
    }
    // 实现自定义的KeyedProcessFunction
    public static class FakeWindowResult extends KeyedProcessFunction<String, Event, String>{
        private Long windowSize; // 窗口大小

        public FakeWindowResult(Long windowSize) {
            this.windowSize = windowSize;
        }
        // 定义状态 MapState, 用来保存每个窗口中统计的count值
        MapState<Long, Long> windowUrlCountMapState;

        @Override
        public void open(Configuration parameters) throws Exception {
            windowUrlCountMapState = getRuntimeContext().getMapState(new MapStateDescriptor<Long, Long>("window-count", Long.class, Long.class));
        }

        @Override
        public void processElement(Event value, KeyedProcessFunction<String, Event, String>.Context ctx, Collector<String> out) throws Exception {
            // 每来一条数据，根据时间戳判断属于哪个窗口（窗口分配器）
            Long windowStart = value.timestamp / windowSize * windowSize;
            Long windowEnd = windowStart + windowSize;
            // 注册 end-1的定时器
            ctx.timerService().registerEventTimeTimer(windowEnd - 1);
            // 更新状态，进行增量聚合
            if (windowUrlCountMapState.contains(windowStart)) {
                Long count = windowUrlCountMapState.get(windowStart);
                windowUrlCountMapState.put(windowStart, count+1);
            }else {
                windowUrlCountMapState.put(windowStart, 1L);
            }
        }
        // 定时器触发时输出计算结果

        @Override
        public void onTimer(long timestamp, KeyedProcessFunction<String, Event, String>.OnTimerContext ctx, Collector<String> out) throws Exception {
            long windowEnd = timestamp + 1;
            long windowStart = windowEnd - windowSize;
            Long count = windowUrlCountMapState.get(windowStart);
            out.collect("窗口" +new Timestamp(windowStart) + " ~ " + new Timestamp(windowEnd) + "   url: " + ctx.getCurrentKey() + "  count: " +count);
            // 模拟窗口关闭， 清除map中对应的key-value
            windowUrlCountMapState.remove(windowStart);
        }
    }

}
